### ISSUE TITLE
`:sh` is sluggish

### ISSUE THREAD

**Issue by chdiza** [macvim-dev/macvim#652]
If I do `:sh`, then there is a noticeable delay between hitting keys and movement on the screen.

This is a thousand times worse when one first does `:set go+=!` to take advantage of the new feature of using the `terminal` feature for drawing rather than the native dumb terminal.  In this case, often the shell prompt doesn't even appear until I start typing.  If I enter some command (inside this shell) and hit RETURN, then once more the prompt won't appear until I start typing again.  This is really unusable, whereas with `&go` not containing "!", it's merely annoying.

**Comment by eirnym** [comment]
Does it still an issue? I can't reproduce it in build 155

**Comment by chdiza** [comment]
Yes.

Now if I do `:set go+=!` (on Mojave anyway) it's no longer *extra* sluggish, but it's still noticeably sluggish.

**Comment by eirnym** [comment]
I see the noticeable delay on entering the terminal in MacVim, but not during working in it. I compare it to Apple Terminal.

I have a bit heavy zshrc to have a good comparison.

**Comment by chdiza** [comment]
> on entering the terminal 

I'm not talking about the `:terminal`.  I'm talking about doing `:sh` in the gui.

The delay is between the pressing of each key (once `:sh` mode has been entered) and the updating of the screen.

**Comment by eirnym** [comment]
If I don't set `:set go+=!` the `:sh` is fast, but doesn't know much as a "terminal" (no colours, no key bindings, etc).
After I've added the settings, it seems, the command calls `:terminal` in the background and I see the same. 

I experience a quite long delay at the beginning (any keypress stops this delay) and then I see no delay between any keypress and draw. 

You can checkout my [.zsh config](https://bitbucket.org/eirnym/.zsh) and see from my point of view. 

I use Mojave/10.14.6 (beta), zsh as a shell and MacVim from GitHub.

**Comment by eirnym** [comment]
My expectations were confirmed that `:sh` runs `:terminal` if `:set go+=!` option has been set. The confirmation you can read in `ex_cmds.c`, `do_shell` function. The very slow transition is because `display_errors()` was called and there's no messages to be shown, so you have to press a key. 

You can fill an issue to the upstream

**Comment by chdiza** [comment]
> You can fill an issue to the upstream

I will not do that, since there is no delay in plain console vim.  The problem is in MacVim, and it happens whether or not "!" is in `&go`.

**Comment by chdiza** [comment]
Also, doing `:terminal` in MacVim produces no laggy input or responses.  It's only when doing `:sh`.

**Comment by ychin** [comment]
Just to make sure I understand:
1. `:sh` is sluggish (meaning there's a long delay between keypress and response), but `:terminal` is not.
2. It doesn't matter whether `&go` has `!` is set.
3. You are using macOS Mojave.

Is that correct?

Do you have CoreText renderer turned on, or any custom set defaults for MacVim? I'm asking as I'm not seeing this issue on my Mojave machine.

**Comment by chdiza** [comment]
1. Hit `:sh<CR>`.  Wait for your shell prompt.  Start typing.  Notice the delay between each subsequent key press and response.  Now do the same thing, except hit `:terminal<CR>` instead of `:sh<CR>`.  Notice there's no sluggishness.

2. For the procedure in (1), there is sluggish whether or not `!` is present in `&go`.  It used to be that the sluggishness was far worse with `!`, but now (at least on Mojave) the sluggishness gap is narrower.

3.  I'm using Mojave right now, but this issue has been around for many, many years.  I reported it in this thread before Mojave was out.  The issue might even have been here for over a decade; I'm pretty sure I remember it on Leopard!

It happens even when started with `-u NONE -U NONE --noplugin` and all that.  It's got nothing to do with my bash `.profile` as far as I can see.  (Hard to see how that would affect individual key latency.)  I'm pretty sure that a while back I moved all that stuff out of the way and the problem is still there.

I do have the CoreText renderer on:
```
$ defaults read -app MacVim
    MMAutosaveColumns = 80;
    MMAutosaveRows = 53;
    MMCurrentPreferencePane = Advanced;
    MMFullScreenFadeTime = 0;
    MMNativeFullScreen = 0;
    MMShareFindPboard = 0;
    MMTextInsetBot = 0;
    MMTextInsetTop = 0;
    MMTopLeftPoint = "{717, 777}";
    MMUntitledWindow = 1;
    NSNavLastRootDirectory = "~";
    NSNavLastUserSetHideExtensionButtonState = 0;
    NSNavPanelExpandedSizeForSaveMode = "{841, 448}";
    NSQuotedKeystrokeBinding = "";
    NSRepeatCountBinding = "";
    "NSWindow Frame NSFontPanel" = "769 110 445 237 0 0 1280 777 ";
    SUEnableAutomaticChecks = 0;
    SUHasLaunchedBefore = 1;
```

I get this behavior now on several different macOS machines, and got it years ago on several others.  E.g.,  MacBookPro9,2; an iMac from 2012; an iMac from 2008; an iMac from 2009; a PowerBookG4 from 2004; a MacBook from 2007.

**Comment by chdiza** [comment]
The lag is particularly noticeable if you type a longish line and then hold down "delete", for example.

I just created a fresh macOS user account and tried it in there, and see the sluggishness.  This means there's no `.profile` or `.vim*` stuff causing it.

**Comment by eirnym** [comment]
> > You can fill an issue to the upstream
> 
> I will not do that, since there is no delay in plain console vim. The problem is in MacVim, and it happens whether or not "!" is in `&go`.

Upstream is not only for teminal Vim, and `do_shell` function and `:terminal` are core functionality. As I've mentioned, it "displays the errors" which are empty.

**Comment by eirnym** [comment]
I see, that terminal in MacVim has it's own delay in comparison to Apple Terminal and same test case. But `:terminal` and `:sh` behave the same with `!` option set (this option in Vim Core code is named `GO_TERMINAL`, open terminal when do `:sh`)

**Comment by chdiza** [comment]
> But :terminal and :sh behave the same with ! option set

Are you reporting the results of testing locally, or are you just saying what should *theoretically* be the case?

> I see, that terminal in MacVim has it's own delay in comparison to Apple Terminal

I don't understand what this means.  Are you saing that after doing `:terminal` in MacVim there is keypress latency that isn't present if you just open up Terminal.app and start typing?  Or are you comparing `:terminal` in MacVim to `:terminal` in console Vim?

> Upstream is not only for teminal Vim, and do_shell function and :terminal are core functionality. As I've mentioned, it "displays the errors" which are empty.

The business about "displays the errors" seems to be completely irrelevant to most of this thread.  I'm not talking about a one-time delay before seeing initial the bash prompt after doing `:sh` or `:terminal`.  That was only one of the problems I raised in this thread, and it was only a secondary complaint; it seems to be gone at least in Mojave.  I'm talking about the first and primary problem mentioned in this thread: a continual lag between each keypress and when it appears on the screen.

**Comment by eirnym** [comment]
> Are you reporting the results of testing locally, or are you just saying what should theoretically be the case?

This is written in Vim's code. file `ex_cmds.c`

> I don't understand what this means. Are you saing that after doing :terminal in MacVim there is keypress latency that isn't present if you just open up Terminal.app and start typing? Or are you comparing :terminal in MacVim to :terminal in console Vim?

Common GUI latency is present, and I haven't tested this with #858 
There's no latency in comparison between `:terminal` and `:sh`. MacVim is _a bit_ (not much) slower in comparison between `Terminal.app` and `MacVim.app`.

**Comment by eirnym** [comment]
#858 gives a good boost for both `:sh` and `:terminal` performance. It's worth to try it.

**Comment by chdiza** [comment]
I have tried it.  I've been using it almost nonstop for the last week or so.

**Comment by ychin** [comment]
I can see the sluggishness now. It honestly doesn't seem terrible with `set go+=!` but it's there, and it's curious that it behaves slightly differently from `:terminal` (easiest way to repro is just to hold down the Return key). This is likely not a one-line fix kind of thing, so I will see if #858 fixes this first. If it doesn't it's like because it's refreshing the whole screen's worth of text, then the next step after #858 will likely be to accelerate the text renderer so the unhappy case (full screen refresh) will still work.

**Comment by ychin** [comment]
I can't seem to repro this anymore. I think #858 fixed it. Let me know if that's not the case @chdiza and we can re-open this issue.
